File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/StandardPacket.java
package com.gmail.marcosav2010.communicator.packet;

import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder;
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder;

import java.io.IOException;

public abstract class StandardPacket extends AbstractPacket {

    @Override
    public boolean isStandard() {
        return true;
    }

    @Override
    public void encode(PacketEncoder out) throws IOException {
        encodeContent(out);
    }

    protected abstract void encodeContent(PacketEncoder out) throws IOException;

    @Override
    public void decode(PacketDecoder in) throws IOException {
        decodeContent(in);
    }

    protected abstract void decodeContent(PacketDecoder in) throws IOException;
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/StandardPacket.kt
package com.gmail.marcosav2010.communicator.packet

import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder

abstract class StandardPacket : AbstractPacket() {

    override val isStandard
        get() = true

    override fun encode(out: PacketEncoder) = encodeContent(out)

    protected abstract fun encodeContent(out: PacketEncoder)

    override fun decode(`in`: PacketDecoder) = decodeContent(`in`)

    protected abstract fun decodeContent(`in`: PacketDecoder)
}